File: python3/fibonacci_number.py
"""
Space   : O(n)
Time    : O(n)
"""


class Solution:
    def fib(self, n: int) -> int:
        mem = [0, 1]
        if n <= 1:
            return mem[n]

        for i in range(2, n+1):
            mem.append(mem[i-1] + mem[i-2])

        return mem[-1]

File: golang/fibonacci_number.go
package golang

// Space   : O(n)
// Time    : O(n)

func fib(n int) int {
    mem := []int{0, 1}
    
    if n <= 1 {
        return mem[n]
    }
    
    for i := 2; i <= n; i++ {
        mem = append(mem, mem[i-1] + mem[i-2])
    }
    
    return mem[n]
}